pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -353,52 +381,97 @@ func (service *SecurityPolicyService) createOrUpdateSecurityPolicy(obj *v1alpha1
+		changed, stale := common.CompareResources(GroupsPtrToComparable(existingProjectGroups), GroupsToComparable(projectGroups))
+		changedProjectGroups, staleProjectGroups := ComparableToGroups(changed), ComparableToGroups(stale)
+		if len(changedProjectGroups) == 0 && len(staleProjectGroups) == 0 {
+			log.Info("project groups are not changed, skip updating them")
+		}
+
+		// 1.Remove stale nsxRules
+		if len(staleProjectGroups) != 0 {
+			// Before deleting the stale project groups and project shares, we must delete the nsxRules to which the project shares are referencing
+			// The deleted stale nsxRules might also include other stale rules to which the project shares are not referencing,
+			// However, we can delete these rules together because the stale rule deletion process is an intermediate step that causes temporary inconsistence
+			// between K8s CR resource and NSX resource. Eventual resource consistency is guaranteed after all the steps below completed.
+			finalSecurityPolicy.Rules = staleRules
+			orgRoot, err := service.WrapHierarchyVpcSecurityPolicy(finalSecurityPolicy, nil, vpcInfo)
+			if err != nil {
+				log.Error(err, "failed to wrap stale rules for project stale shares")
+				return err
 			}
+			err = service.NSXClient.OrgRootClient.Patch(*orgRoot, &EnforceRevisionCheckParam)
+			if err != nil {
+				log.Error(err, "failed to delete rules for project stale shares")
+				return err
+			}
+			// Add changed nsxRules into SecurityPolicy which will be handled later
+			finalSecurityPolicy.Rules = changedRules
+			for i := len(staleProjectGroups) - 1; i >= 0; i-- {
+				staleProjectGroups[i].MarkedForDelete = &MarkedForDelete
+			}
+		}
 
-			// 1.Create/update project level groups
-			finalProjectGroups, err = service.createOrUpdateProjectGroups(obj, projectGroups, createdFor)
+		finalProjectGroups = append(finalProjectGroups, staleProjectGroups...)
+		finalProjectGroups = append(finalProjectGroups, changedProjectGroups...)
+		existingProjectShares := shareStore.GetByIndex(common.TagScopeSecurityPolicyCRUID, string(obj.UID))
+		changed, stale = common.CompareResources(SharesPtrToComparable(existingProjectShares), SharesToComparable(projectShares))
+		changedProjectShares, staleProjectShares := ComparableToShares(changed), ComparableToShares(stale)
+		if len(changedProjectShares) == 0 && len(staleProjectShares) == 0 {
+			log.Info("project shares are not changed, skip updating them")
+		}
+		for i := len(staleProjectShares) - 1; i >= 0; i-- {
+			staleProjectShares[i].MarkedForDelete = &MarkedForDelete
+		}
+		finalProjectShares = append(finalProjectShares, staleProjectShares...)
+		finalProjectShares = append(finalProjectShares, changedProjectShares...)
+
+		// 2.Create/update project groups and shares
+		if len(finalProjectGroups) != 0 || len(finalProjectShares) != 0 {
+			projectInfra, err := service.WrapHierarchyProjectResources(finalProjectShares, finalProjectGroups)
```

Comment:
Can we wrap securitypolicy & projectshares & project groups into one object?

Reply:
Unfortunately, we can not warp securitypolicy & project shares & project groups into obj, because securitypolicy is put in VPC level, so we need to use OrgRootClient to wrap securitypolicy, rules and groups of VPC level into one API Obj. As for project shares & project groups, which are put in project level, we use ProjectInfraClient to wrap these two into one API obj.

Reply:
Much appreciation for bringing up this critical point for discussion. Ultimately, we have find a solution that we can wrap ecuritypolicy, rules, groups in VPC level & projectshares & project groups into one Hierarchy tree object. This approach  help us to manipulate all sort of objects in one API call, which absolutely helps the code quality in term of performance and security.  The only words I want to say, Thank you.